### PR TITLE
Add onnxruntime

### PR DESCRIPTION
**Pull request by hqucms**
As discussed in https://github.com/cms-sw/cmssw/issues/27458.

@davidlange6 This builds only the C API -- but the python package might also be helpful if analyzers want to run some ML inference in their python-based analyses.

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

(the cmake update is indeed required - that will rebuild lots of things I suspect)

File: onnxruntime.spec [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+   -Donnxruntime_ENABLE_MICROSOFT_INTERNAL=OFF \
+   -Donnxruntime_USE_BRAINSLICE=OFF \
+   -Donnxruntime_USE_NUPHAR=OFF \
+   -Donnxruntime_USE_EIGEN_THREADPOOL=OFF \
+   -Donnxruntime_USE_TENSORRT=OFF \
+   -Donnxruntime_CROSS_COMPILING=OFF \
+   -Donnxruntime_USE_FULL_PROTOBUF=OFF \
+   -Donnxruntime_DISABLE_CONTRIB_OPS=OFF \
+   -Donnxruntime_BUILD_UNIT_TESTS=OFF \
+   ../cmake
```
Comment (slava77) [inline onnxruntime.spec]: I do not see any specifics for the build architecture.
What does this default to? We still need to restrict to SSE4 (or is it SSE3?)

Comment (hqucms) [reply onnxruntime.spec]: It seems that it supports dynamic architecture -- I built it once on a machine with AVX and it runs fine on another machine without it.

**Comment by hqucms** [comment]
I cannot reproduce the error -- igprof builds fine with the cmake update on my side.

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
@hqucms , the build error is this. May be you have unwind installed on the system that is why cmake did not fail.

```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
UNWIND_INCLUDE_DIR
```

**Comment by smuzaffar** [comment]
During PR tests we create symlink for packages which are not changed. In this case we create symlink for unwwind. I think problem is here
https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_0_X/gcc700/igprof.spec#L16
which just rsync symlink instead of contents of unwind and atoim_ops. I am looking in to this now.

**Comment by hqucms** [comment]
> During PR tests we create symlink for packages which are not changed. In this case we create symlink for unwwind. I think problem is here
> https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_0_X/gcc700/igprof.spec#L16
> which just rsync symlink instead of contents of unwind and atoim_ops. I am looking in to this now.

@smuzaffar Thank you for looking into this! Indeed when I built igprof, unwind was downloaded rather than symlinked, which is probably why it built successfully on my machine.

**Comment by smuzaffar** [comment]
please test

**Comment by hqucms** [comment]
@smuzaffar What's the actual error? I suspect that we need to add `zlib python3` to `BuildRequires` -- they were pointing to system libs in my environment.

**Comment by smuzaffar** [comment]
@hqucms , to see the actual error , you can go to  https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c31479/1386/summary.html  and the click on the "External Build Logs" link and then go to external -> onnxruntime -> 0.4.0 -> log
Yes looks like( https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c31479/1386/external/onnxruntime/0.4.0/log ) it did pick up some things from system. 
By the way, does it only build for python3? 

```
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.7") 
-- Found PNG: /usr/lib64/libpng.so (found version "1.5.13") 
CMake Error at /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc700/external/cmake/3.14.5/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find PythonInterp: Found unsuitable version "2.7.14", but
  required is at least "3.4" (found
  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc700/external/python/2.7.14-pafccj/bin/python)
Call Stack (most recent call first):
  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc700/external/cmake/3.14.5/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:376 (_FPHSA_FAILURE_MESSAGE)
  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc700/external/cmake/3.14.5/share/cmake-3.14/Modules/FindPythonInterp.cmake:160 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:205 (find_package)

```

**Comment by davidlange6** [comment]
onnx-runtime only supports python3. That should be fine but it means having python3 in the spec rather than python (ah - but python is not there at all yet)

**Comment by smuzaffar** [comment]
@hqucms , I also have updated the spec to use correct URL for source. We also need to make sure that the eigen, protobuf and other submodules used by onnxruntime are competible with the one used by CMS

**Comment by davidlange6** [comment]
actually it likely is a problem if this is using python behind the scenes in the c(++) api. 

(but btw - can we do the cmake update separately to make these iterations faster?)

**Comment by hqucms** [comment]
@smuzaffar Thank you very much for pointing me to the log files and for updating the spec (the `submodules` option is really nice to have)!

Regarding this:
> We also need to make sure that the eigen, protobuf and other submodules used by onnxruntime are competible with the one used by CMS

protobuf and eigen are not needed at run time (below is the the output of `ldd`). Could there still be compatibility issue then?

```
> ldd libonnxruntime.so 
        linux-vdso.so.1 =>  (0x00007ffd8c7ab000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f8999f23000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f8999d07000)
        libstdc++.so.6 => /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-pafccj/lib64/libstdc++.so.6 (0x00007f899a81d000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f8999a05000)
        libgcc_s.so.1 => /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-pafccj/lib64/libgcc_s.so.1 (0x00007f899a804000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f8999638000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f899a7b0000)
```

**Comment by davidlange6** [comment]
ok, it seems major surgery is needed to match the treatment of externals with the reset of the cmsdist stack

https://github.com/microsoft/onnxruntime/tree/master/cmake/external

**Comment by hqucms** [comment]
> actually it likely is a problem if this is using python behind the scenes in the c(++) api. 

I think it doesn't call python in the C API -- it needs python at compilation time probably just for this:
https://github.com/microsoft/onnxruntime/blob/master/cmake/onnxruntime.cmake#L23.

> (but btw - can we do the cmake update separately to make these iterations faster?)

Good idea! Shall I do it or better leave it for @smuzaffar?

**Comment by smuzaffar** [comment]
https://github.com/cms-sw/cmsdist/pull/5087 should update cmake.

**Comment by hqucms** [comment]
> #5087 should update cmake.

Thanks @smuzaffar! I updated the spec to explicitly use the `python3` executable in the CMS software stack, hopefully it builds fine this time.

**Comment by mrodozov** [comment]
please test
lets see

**Comment by smuzaffar** [comment]
please test
just to have fresh results

**Comment by smuzaffar** [comment]
+externals

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
+externals

**Comment by smuzaffar** [comment]
@hqucms , this is now available in CMSSW_11_0_X IBs. Note that occxruntime currently does not support PowerPC so that is why it is not available for ppc64le SCRAM ARCH IBs.

**Comment by hqucms** [comment]
@smuzaffar Thank you very much! I will test occxruntime from the IBs and proceed from there.

**Pull request merged**